redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt

```diff
@@ -26,101 +26,197 @@ import app.cash.redwood.layout.api.Constraint
+import platform.UIKit.UICollectionViewLayout
+import platform.UIKit.UICollectionViewScrollDirection
 import platform.UIKit.UIControlEventValueChanged
+import platform.UIKit.UIEdgeInsetsMake
+import platform.UIKit.UILabel
 import platform.UIKit.UIRefreshControl
-import platform.UIKit.UITableView
-import platform.UIKit.UITableViewCell
-import platform.UIKit.UITableViewCellStyle
-import platform.UIKit.UITableViewCellStyle.UITableViewCellStyleDefault
-import platform.UIKit.UITableViewDataSourceProtocol
-import platform.UIKit.UITableViewDelegateProtocol
+import platform.UIKit.UIScrollView
 import platform.UIKit.UIView
+import platform.UIKit.indexPathForRow
 import platform.UIKit.item
 import platform.darwin.NSInteger
 import platform.darwin.NSObject
 
-private const val reuseIdentifier = "cell"
+internal interface ViewPortItems: Widget.Children<UIView> {
+  var placeholders: MutableList<Widget<UIView>>
+  var itemsBefore: Int
+  var itemsAfter: Int
+  fun itemForGlobalIndex(index: Int): UIView
+  fun itemCount(): Int
+}
+
+internal open class UIViewLazyList() : LazyList<UIView>, ChangeListener {
 
-internal open class UIViewLazyList() : LazyList<UIView> {
+  override val items: ViewPortItems = object : ViewPortItems {
+    override var placeholders = mutableListOf<Widget<UIView>>()
 
-  private val itemsList = mutableListOf<Widget<UIView>>()
+    override var itemsBefore: Int = 0
+    override var itemsAfter: Int = 0
 
-  override val placeholder: Widget.Children<UIView> = MutableListChildren()
+    private val viewPortList = mutableListOf<Widget<UIView>>()
 
-  override val items: Widget.Children<UIView> = object : Widget.Children<UIView> {
     override fun insert(index: Int, widget: Widget<UIView>) {
-      itemsList.add(index, widget)
-      tableView.reloadData()
+      viewPortList.add(index, widget)
     }
 
     override fun move(fromIndex: Int, toIndex: Int, count: Int) {
-      itemsList.move(fromIndex, toIndex, count)
-      tableView.reloadData()
+      viewPortList.move(fromIndex, toIndex, count)
     }
 
     override fun remove(index: Int, count: Int) {
-      itemsList.remove(index, count)
-      tableView.reloadData()
+      viewPortList.remove(index, count = count)
+    }
+
+    override fun onModifierUpdated() {}
+
+    // Fetch the item from the viewPortList relative to the entire collection view
+    override fun itemForGlobalIndex(index: Int): UIView {
+      val viewPortIndex: Int = max(index - itemsBefore, 0)
+
+      viewPortList.getOrNull(viewPortIndex)?.let {
+        return it.value
+      }
+
+      // If we don't have a value, fallback to our pools of placeholders
+      val placeholderIndex = index % placeholders.size
+      return placeholders.get(placeholderIndex).value
     }
 
-    override fun onModifierUpdated() {
+    override fun itemCount(): Int {
+      // TODO: replace with this with "ItemCounts"
+      return max(itemsBefore - 1, 0 ) + viewPortList.count() + itemsAfter
```

Comment:
I'm curious about the `max(itemsBefore - 1, 0 )` bit. Why do we want 1 less than `itemsBefore`?

Essentially, I would've thought it'd be:

```suggestion
      return itemsBefore + viewPortList.count() + itemsAfter
```

Reply:
i think i was seeing an off by 1 error, but i think i got that mixed up with the index. This will get replaced with the ItemCounts anyways